util/dockerutil/progress.go

```diff
@@ -1,47 +1,79 @@
 package dockerutil
 
 import (
-	"errors"
+	"encoding/json"
 	"io"
 	"time"
 
 	"github.com/docker/buildx/util/progress"
-	"github.com/docker/cli/cli/streams"
 	"github.com/docker/docker/pkg/jsonmessage"
 	"github.com/moby/buildkit/client"
-	"github.com/moby/buildkit/identity"
-	"github.com/opencontainers/go-digest"
 )
 
-func fromReader(w progress.Writer, name string, rc io.ReadCloser) error {
-	dgst := digest.FromBytes([]byte(identity.NewID()))
-	tm := time.Now()
+const minProgressDelta = 0.50 // %
 
-	vtx := client.Vertex{
-		Digest:  dgst,
-		Name:    name,
-		Started: &tm,
-	}
+func fromReader(l progress.SubLogger, rc io.ReadCloser) error {
+	started := map[string]client.VertexStatus{}
 
-	w.Write(&client.SolveStatus{
-		Vertexes: []*client.Vertex{&vtx},
-	})
+	defer func() {
+		for _, st := range started {
+			st := st
+			if st.Completed == nil {
+				now := time.Now()
+				st.Completed = &now
+				l.SetStatus(&st)
+			}
+		}
+	}()
 
-	err := jsonmessage.DisplayJSONMessagesToStream(rc, streams.NewOut(io.Discard), nil)
-	if err != nil {
-		if jerr, ok := err.(*jsonmessage.JSONError); ok {
-			err = errors.New(jerr.Message)
+	dec := json.NewDecoder(rc)
+	var parsedErr error
+	var jm jsonmessage.JSONMessage
+	for {
+		if err := dec.Decode(&jm); err != nil {
+			if parsedErr != nil {
+				return parsedErr
+			}
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+		if jm.Error != nil {
+			parsedErr = jm.Error
+		}
+		if jm.ID == "" {
+			continue
+		}
+		id := "loading layer " + jm.ID
+		st, ok := started[id]
+		if !ok {
+			if jm.Progress != nil {
+				now := time.Now()
+				st = client.VertexStatus{
+					ID:      id,
+					Started: &now,
+				}
+			} else {
+				continue
+			}
 		}
+		st.Timestamp = time.Now()
+		var progressDelta float64
+		if jm.Progress != nil && jm.Progress.Total > 0 {
+			progressDelta = float64(jm.Progress.Current-st.Current) / float64(jm.Progress.Total)
+		}
+		if jm.Progress != nil && jm.Status == "Loading layer" && (st.Current == 0 || progressDelta >= minProgressDelta) {
+			st.Current = jm.Progress.Current
+			st.Total = jm.Progress.Total
+		}
```

Comment:
@tonistiigi As discussed, increased minimal progress delta to reduce log output (PR desc updated with logs).

Previously:

```
#35 loading layer 5e7ac2b0f7d2 5.31MB / 27.73MB 0.1s
#35 loading layer 5e7ac2b0f7d2 12.98MB / 27.73MB 0.2s
#35 loading layer 5e7ac2b0f7d2 20.64MB / 27.73MB 0.3s
#35 loading layer 5e7ac2b0f7d2 27.73MB / 27.73MB 0.4s
#35 loading layer 6b4ce6ad2a03 557.06kB / 70.77MB
#35 loading layer 6b4ce6ad2a03 17.83MB / 70.77MB 0.2s
#35 loading layer 6b4ce6ad2a03 26.74MB / 70.77MB 0.3s
#35 loading layer 6b4ce6ad2a03 38.44MB / 70.77MB 0.4s
#35 loading layer 6b4ce6ad2a03 46.24MB / 70.77MB 0.5s
#35 loading layer 6b4ce6ad2a03 56.26MB / 70.77MB 0.6s
#35 loading layer 6b4ce6ad2a03 63.50MB / 70.77MB 0.7s
#35 loading layer 6b4ce6ad2a03 70.77MB / 70.77MB 0.8s
#35 loading layer ecfe8dc499b5 163.84kB / 13.78MB
#35 loading layer ecfe8dc499b5 13.78MB / 13.78MB 0.1s
#35 loading layer 2fea51b45f4f 557.06kB / 52.93MB
#35 loading layer 2fea51b45f4f 10.03MB / 52.93MB 0.1s
#35 loading layer 2fea51b45f4f 19.50MB / 52.93MB 0.2s
#35 loading layer 2fea51b45f4f 31.20MB / 52.93MB 0.3s
#35 loading layer 2fea51b45f4f 39.55MB / 52.93MB 0.4s
#35 loading layer 2fea51b45f4f 52.93MB / 52.93MB 0.5s
#35 loading layer 5f70bf18a086 32B / 32B
#35 loading layer 992b18fde4b2 8.26MB / 17.46MB 0.1s
#35 loading layer 992b18fde4b2 15.53MB / 17.46MB 0.2s
#35 loading layer 992b18fde4b2 17.46MB / 17.46MB 0.2s
#35 loading layer 9c87ca044b96 196.61kB / 17.77MB
#35 loading layer 9c87ca044b96 15.93MB / 17.77MB 0.2s
#35 loading layer 9c87ca044b96 17.77MB / 17.77MB 0.2s
#35 loading layer 17fde1d06fde 458.75kB / 45.74MB
#35 loading layer 17fde1d06fde 27.98MB / 45.74MB 0.2s
#35 loading layer 17fde1d06fde 45.74MB / 45.74MB 0.3s
#35 loading layer 956dbc1a6d25 332B / 332B
#35 loading layer 956852dd3d1b 116B / 116B
#35 loading layer 7eaadc8c5cc9 556B / 556B
#35 loading layer c0e431c2bc5e 1.03kB / 1.03kB 0.0s
#35 loading layer 5e7ac2b0f7d2 27.73MB / 27.73MB 3.0s done
#35 loading layer 6b4ce6ad2a03 70.77MB / 70.77MB 2.5s done
#35 loading layer ecfe8dc499b5 13.78MB / 13.78MB 1.7s done
#35 loading layer 2fea51b45f4f 52.93MB / 52.93MB 1.6s done
#35 loading layer 5f70bf18a086 32B / 32B 1.0s done
#35 loading layer 992b18fde4b2 17.46MB / 17.46MB 1.0s done
#35 loading layer 9c87ca044b96 17.77MB / 17.77MB 0.8s done
#35 loading layer 17fde1d06fde 45.74MB / 45.74MB 0.5s done
#35 loading layer 956dbc1a6d25 332B / 332B 0.1s done
#35 loading layer 956852dd3d1b 116B / 116B 0.1s done
#35 loading layer 7eaadc8c5cc9 556B / 556B 0.1s done
#35 loading layer c0e431c2bc5e 1.03kB / 1.03kB 0.1s done
```

Now:

```
#36 loading layer 425060b76f85 294.91kB / 27.73MB 0.1s
#36 loading layer 425060b76f85 20.64MB / 27.73MB 0.3s
#36 loading layer 9322ccb2ce56 557.06kB / 70.77MB
#36 loading layer 9322ccb2ce56 46.24MB / 70.77MB 0.5s
#36 loading layer 2ec616245a7b 163.84kB / 13.78MB
#36 loading layer 2ec616245a7b 12.12MB / 13.78MB 0.1s
#36 loading layer 40b1e4cd99b0 557.06kB / 52.93MB
#36 loading layer 40b1e4cd99b0 39.55MB / 52.93MB 0.4s
#36 loading layer 5f70bf18a086 32B / 32B
#36 loading layer 1b2dee13b9a1 196.61kB / 17.46MB 0.1s
#36 loading layer 1b2dee13b9a1 15.34MB / 17.46MB 0.2s
#36 loading layer b4c8beeaca6b 196.61kB / 17.77MB
#36 loading layer b4c8beeaca6b 15.93MB / 17.77MB 0.2s
#36 loading layer 3aabf2e36740 458.75kB / 45.74MB
#36 loading layer 3aabf2e36740 27.98MB / 45.74MB 0.2s
#36 loading layer 80022eaf27ce 333B / 333B
#36 loading layer 7e0cb0eb0de5 116B / 116B
#36 loading layer 71e71eda1857 553B / 553B
#36 loading layer 425060b76f85 20.64MB / 27.73MB 3.0s done
#36 loading layer 9322ccb2ce56 46.24MB / 70.77MB 2.5s done
#36 loading layer 2ec616245a7b 12.12MB / 13.78MB 1.7s done
#36 loading layer 40b1e4cd99b0 39.55MB / 52.93MB 1.6s done
#36 loading layer 5f70bf18a086 32B / 32B 1.0s done
#36 loading layer 1b2dee13b9a1 15.34MB / 17.46MB 1.0s done
#36 loading layer b4c8beeaca6b 15.93MB / 17.77MB 0.8s done
#36 loading layer 3aabf2e36740 27.98MB / 45.74MB 0.5s done
#36 loading layer 80022eaf27ce 333B / 333B 0.1s done
#36 loading layer 7e0cb0eb0de5 116B / 116B 0.1s done
#36 loading layer 71e71eda1857 553B / 553B 0.1s done
#36 loading layer 51545e2a7115 1.03kB / 1.03kB 0.1s done
```